File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraysWithKDistinct = function(nums, k) {
    return atMostK(k) - atMostK(k-1)
    
    function atMostK(k) {
        let res = 0
        let i = 0
        const map = {}
        for (let j=0; j < nums.length; j++) {
            if (!(nums[j] in map) || (map[nums[j]] === 0)) k--
            map[nums[j]] = map[nums[j]] + 1 || 1
            while (k < 0) {
                map[nums[i]]--
                if (map[nums[i]] === 0) k++
                i++
            }
            res += j - i + 1
        }
        return res
    }
};

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        
        def helper(k: int) -> int: 
            i, res, mp = 0, 0, {}
            
            for j, num in enumerate(nums):
                if mp.get(num, 0) == 0: k -= 1
                mp[num] = mp.get(num, 0) + 1
                
                while k < 0:
                    mp[nums[i]] -= 1
                    if mp[nums[i]] == 0: k += 1
                    i += 1
                    
                res += j - i + 1
                
            return res             
        
        return helper(k) - helper(k-1)